### PR TITLE
Hughsie/fu usb descriptor

### PR DESCRIPTION
**Pull request by hughsie**

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 I'm done for today, review very welcome. Thanks!

File: libfwupdplugin/fu-usb-hid-descriptor.c [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	object_class->finalize = fu_usb_hid_descriptor_finalize;
+	firmware_class->parse = fu_usb_hid_descriptor_parse;
+}
+
+/**
+ * fu_usb_hid_descriptor_new:
+ *
+ * Return value: a new #FuUsbHidDescriptor object.
+ *
+ * Since: 2.0.0
```
Comment (superm1) [inline libfwupdplugin/fu-usb-hid-descriptor.c:202]: 2.0.2

Comment (hughsie) [reply libfwupdplugin/fu-usb-hid-descriptor.c:202]: An entire days work for that comment! :)

Comment (superm1) [reply libfwupdplugin/fu-usb-hid-descriptor.c]: Heh, I didn't finish looking through the rest yet.  Juggling a bunch, I'll let you know if anything else stands out ;P

File: libfwupdplugin/fu-efi-file.c [hunk 2]
```diff
@@ -241,11 +241,13 @@ fu_efi_file_write(FuFirmware *firmware, GError **error)
 	blob = fu_efi_file_write_sections(firmware, error);
 	if (blob == NULL)
 		return NULL;
-	if (!fwupd_guid_from_string(fu_firmware_get_id(firmware),
-				    &guid,
-				    FWUPD_GUID_FLAG_MIXED_ENDIAN,
-				    error))
-		return NULL;
+	if (fu_firmware_get_id(firmware) != NULL) {
```
Comment (superm1) [inline libfwupdplugin/fu-efi-file.c:244]: isn't this a programmer error?  Should it be noisier?

Comment (hughsie) [reply libfwupdplugin/fu-efi-file.c:244]: I think in this case a NULL GUID is fine, but point heeded.

File: plugins/bcm57xx/fu-bcm57xx-device.c [hunk 3]
```diff
@@ -52,6 +51,66 @@ fu_bcm57xx_device_probe(FuDevice *device, GError **error)
...
+
+	/* save */
+	if (fu_context_has_flag(fu_device_get_context(FU_DEVICE(self)),
+				FU_CONTEXT_FLAG_SAVE_EVENTS)) {
+		event = fu_device_save_event(FU_DEVICE(self), event_id);
+		fu_device_event_set_data(event, "Data", buf, bufsz);
+	}
+
+	/* we can't use the emulation support in fu_udev_device_ioctl() as
+	 * the buffer is specified indirectly using ifr_data */
```
Comment (superm1) [inline plugins/bcm57xx/fu-bcm57xx-device.c:92]: I _think_ this is the same problem with why amdgpu probing can't work either in emulation currently.

File: plugins/dell-dock/fu-dell-dock-status.c [hunk 4]
```diff
@@ -103,12 +103,20 @@ fu_dell_dock_status_open(FuDevice *device, GError **error)
 	if (fu_device_get_proxy(device) == NULL)
 		fu_device_set_proxy(device, fu_device_get_parent(device));
 
+	if (fu_device_get_proxy(device) == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "no proxy");
+		return FALSE;
+	}
```
Comment (superm1) [inline plugins/dell-dock/fu-dell-dock-status.c:109]: wait isn't this impossible?  Look at the check above.

File: plugins/dell-dock/fu-dell-dock-status.c [hunk 5]
```diff
@@ -103,12 +103,20 @@ fu_dell_dock_status_open(FuDevice *device, GError **error)
...
 	return fu_device_open(fu_device_get_proxy(device), error);
 }
 
 static gboolean
 fu_dell_dock_status_close(FuDevice *device, GError **error)
 {
+	if (fu_device_get_proxy(device) == NULL) {
+		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "no proxy");
+		return FALSE;
+	}
```
Comment (superm1) [inline plugins/dell-dock/fu-dell-dock-status.c:119]: likewise, I think this is also impossible.  You wouldn't be able to open if the proxy was NULL.

File: plugins/dell-dock/fu-dell-dock-status.c (hunk 4 above)
Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-status.c:109]: It's kinda unclear, but the logic is sound -- e.g.

	if (fu_device_get_proxy(device) == NULL)
		fu_device_set_proxy(device, fu_device_get_parent(device));
   //then we check fu_device_get_proxy(device)

It's certainly a bit weird. I'll fix it up.

File: plugins/dell-dock/fu-dell-dock-status.c (hunk 5 above)
Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-status.c:119]: Nah, we close() even if open() failed

Comment (superm1) [reply plugins/dell-dock/fu-dell-dock-status.c:119]: While changing that plug-in, do you think we could add emulation data now?

Comment (superm1) [reply plugins/dell-dock/fu-dell-dock-status.c:119]: While changing that plug-in, do you think we could add emulation data now?

Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-status.c:119]: Yup, I've got a physical printed off list of plugins and I'm working my way through them. It takes fricken ages, but it's 100% worthwhile as it's finding real bugs too.

File: plugins/dell-dock/fu-dell-dock-status.c (hunk 4 above)
Comment (hughsie) [reply plugins/dell-dock/fu-dell-dock-status.c:109]: Fixed.

File: plugins/bcm57xx/fu-bcm57xx-device.c (hunk 3 above)
Comment (hughsie) [reply plugins/bcm57xx/fu-bcm57xx-device.c:92]: Ya, the return_pointer stuff will need to be handled in the same way -- but there are other places it's going to fail too; e.g. using g_file_tests on random sysfs files -- we should add a fu_udev_device helper for that so we can emulate it. That'd work for bcm57xx too. The other pain point is amdgpu_device_initialize() but we could wrap that too. I've got an early AMD thinkpad here; goes that have anything new enough for pspvbflash or perhaps you could send me something PVTish that I could use to help?

Comment (hughsie) [reply plugins/bcm57xx/fu-bcm57xx-device.c:92]: 1/3: https://github.com/fwupd/fwupd/pull/7972

**Review by superm1 — APPROVED** [review]

**Pull request merged**